### PR TITLE
Dataloading utils

### PR DESCRIPTION
**Pull request by daler3**
## Description
Work in progress pull request for dataloading utils, dataloaders and datasets.

## Affected Dependencies
Currently using PySyft 2.0. To be changed to not using PySyft at all, or eventually PySyft 3.0

## How has this been tested?
Manually, unit and integration tests to be properly added

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

**Comment by daler3** [comment]
To Resolve #81

File: examples/dualheaded_datautils/dataloaders.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from torch.utils.data._utils.collate import default_collate
+from typing import List, Tuple
+from uuid import UUID
+from uuid import uuid4
+from torch.utils.data import SequentialSampler, RandomSampler, BatchSampler
+
+import datasets
+
+
+"""I think this is not needed anymore"""
```
Comment (TTitcombe) [inline examples/dualheaded_datautils/dataloaders.py:15]: Do you mean we don't need the partitioned dataloader?

File: examples/dualheaded_datautils/datasets.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        self.values_dic = {}
+        for i, l in zip(ids, values):
+            self.values_dic[i] = l
+        self.is_labels = is_labels
+
+        self.ids = torch.Tensor(ids)
+        self.values = torch.Tensor(values) if is_labels else torch.stack(values)
+
+        self.worker_id = None 
+        if worker_id != None:
```
Comment (TTitcombe) [inline examples/dualheaded_datautils/datasets.py:24]: that can simplify to `if worker_id:`

File: examples/dualheaded_datautils/datasets.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    def __len__(self):
+        return len(self.indices)
+
+    def __repr__(self):
+
+        fmt_str = "FederatedDataset\n"
+        fmt_str += f"    Distributed accross: {', '.join(str(x) for x in self.workers)}\n"
+        fmt_str += f"    Number of datapoints: {self.__len__()}\n"
+        return fmt_str
```
Comment (TTitcombe) [inline examples/dualheaded_datautils/datasets.py:151]: newline at the end of the file

File: examples/dualheaded_datautils/enhancedSplitWorkers.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+class EnhanchedWorker():
+    """Single worker with a role (label / data holder) and a model"""
+
+    def __init__(self, worker, dataset, model, level=1):
+        
+        self.worker = worker
+        self.dataset = dataset #It can also be None, and then it would be only computational
+        self.model = model 
+
+        self.level = level if level >= 0 else 0 #it should start from zero, otherwise throw error #TODO: implement error throwing
```
Comment (TTitcombe) [inline examples/dualheaded_datautils/enhancedSplitWorkers.py:27]: simplify to `max(level, 0)`

File: examples/dualheaded_datautils/utils.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import datasets
+from datasets import *
+
+"""
+Utility functions to split and distribute the data across different workers, 
+create vertical datasets and federate them. It also contains datasets and dataloader classes. 
+This code is meant to be used with dual-headed Neural Networks, where there are a bunch of different workers, 
+which agrees on the labels, and there is a server  with the labels only. 
+Code built upon: 
+- Abbas Ismail's (@abbas5253) work on dual-headed NN. In particular, check Configuration 1:
```
Comment (TTitcombe) [inline examples/dualheaded_datautils/utils.py:22]: Does this PR require abbas' PR to be merged?

File: examples/dualheaded_datautils/utils.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        n_workers(optional, default=2): The number of workers to split the data across. If worker_list is not passed, this is necessary to create the split.
+        label_server (optional): the server which owns only the labels (e.g. in a dual-headed NN setting)
+        #TODO: add the code to send labels to the server
+    Returns: 
+        a dictionary holding as keys the workers passed as parameters, or integers corresponding to the split, 
+        and as values a list of lists, where the first element are the single tensor of the data, the second the labels, 
+        the third the index, which is to keep track of the same data point. 
+    """    
+        
+    if worker_list == None:
```
Comment (TTitcombe) [inline examples/dualheaded_datautils/utils.py:55]: `if worker_list:` or `if worker_list is None:`

**Review by TTitcombe — CHANGES_REQUESTED** [review]

File: examples/dualheaded_datautils/dataloaders.py (hunk 1 above)
Comment (daler3) [reply examples/dualheaded_datautils/dataloaders.py:15]: No, I mean that the default pytorch dataloader in PyTorch works, so we do not need a custom one (for how it is done now). See the notebook for an example.
